src/Agent.Sdk/LogPlugin.cs

```diff
@@ -0,0 +1,534 @@
+            {
+                Task memoryUsageMonitor = StartMemoryUsageMonitor(monitorSource.Token);
+
+                Dictionary<string, Task> processTasks = new Dictionary<string, Task>();
+                foreach (var plugin in _plugins)
+                {
+                    // start process plugins background
+                    _trace.Trace($"Start process task for plugin '{plugin.FriendlyName}'");
+                    var task = RunAsync(plugin, tokenSource.Token);
+                    processTasks[plugin.FriendlyName] = task;
+                }
+
+                // waiting for job finish event
+                await _jobFinished.Task;
+                tokenSource.Cancel();
+
+                _trace.Trace($"Wait for all plugins finish process outputs.");
+                foreach (var task in processTasks)
+                {
+                    try
+                    {
+                        await task.Value;
+                        _trace.Trace($"Plugin '{task.Key}' finished log process.");
+                    }
+                    catch (Exception ex)
+                    {
+                        _trace.Output($"Plugin '{task.Key}' failed with: {ex}");
+                    }
+                }
+
+                // Stop monitor
+                monitorSource.Cancel();
+                await memoryUsageMonitor;
+
+                // job has finished, all log plugins should start their finalize process
+                Dictionary<string, Task> finalizeTasks = new Dictionary<string, Task>();
+                foreach (var plugin in _plugins)
+                {
+                    string typeName = plugin.GetType().FullName;
+                    if (!_shortCircuited[typeName].Task.IsCompleted)
+                    {
+                        _trace.Trace($"Start finalize for plugin '{plugin.FriendlyName}'");
+                        var finalize = plugin.FinalizeAsync(_pluginContexts[typeName]);
+                        finalizeTasks[plugin.FriendlyName] = finalize;
+                    }
+                    else
+                    {
+                        _trace.Trace($"Skip finalize for short circuited plugin '{plugin.FriendlyName}'");
+                    }
+                }
+
+                _trace.Trace($"Wait for all plugins finish finalization.");
+                foreach (var task in finalizeTasks)
+                {
+                    try
+                    {
+                        await task.Value;
+                        _trace.Trace($"Plugin '{task.Key}' finished job finalize.");
+                    }
+                    catch (Exception ex)
+                    {
+                        _trace.Output($"Plugin '{task.Key}' failed with: {ex}");
+                    }
+                }
+
+                _trace.Trace($"All plugins finished finalization.");
+            }
+        }
+
+        public void EnqueueOutput(string output)
+        {
+            if (!string.IsNullOrEmpty(output))
```

Comment:
I'm curious on why EnqueueOutput is not passing in empty lines to the plugins. Is this a design choice?

Reply:
good question, i am not sure.
on one side, when plugin is going to parse the line, empty line seems useless.
however, on the other side, empty line might be part of output structure.

Reply:
Right, it might be that that there could be some programs out there that separate sections of the output by empty lines. Removing these empty lines would mean that such sections would be unintentionally stitched together.

That said, the only scenario that I think this matters is the 'save private log' scenario, so its scope its pretty minimal.

Reply:
i will change to `if(output != null)`